src/solver/optimize.cpp

```diff
@@ -1,14 +1,296 @@
+#include <math.h>
+#include <string>  
+#include <iostream> 
+#include <sstream> 
 #include "../components/token.hpp"
 #include "../components/order.hpp"
 #include "../components/amm.hpp"
 #include "./optimize.hpp"
+#include "../../external_libs/json.hpp"
 
-void solve_auction(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms)
+
+// auxiliary function (used temporarily) for computing x^n using the boost library data types
+boost::multiprecision::mpf_float custom_pow(boost::multiprecision::mpz_int x, int n)
+{
+    int i;
+    boost::multiprecision::mpf_float res = 1;
+    int m = (n > 0) ? n : (-n);
+
+    for (i = 0; i < m; i++)
+        res = res * x;
+    if (n > 0)
+        return res;
+    else
+        return (1/res);
+}
+
+// main function that processes a sell order, checks if it can be matched against a single liquidity pool
+// and if so, it also checks whether it gives a strictly better objective
+void process_sell_order(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, int i, ExecutedOrder &best_order)
 {
-    // TO BE COMPLETED
-    int *vertices;
-    int **edges;
+   int j, num_amms;
+    boost::multiprecision::mpz_int x0, y0, x1, y1, x_, y_, x;
```

Comment:
In c++ we usually declare the variables as close as possible to their use. For example x0 could be declared at line 43.

Reply:
I suspect that the compiler in any case takes care of all these, but isn't technically "more correct" to define these variables up top since they are reused throughout all the iterations of the for-loop? Again, I suspect the compiler would do the right thing and not dynamically allocate the variables at the beginning of each iteration, but this is a more explicit way of doing so. I don't have a strong opinion on that so I could change it of course

Reply:
You're right, the compiler will likely move it around, but remember that in c++ creating and destroying an object also runs code. So the place where you declare the object is important, as the constructor and destructor will run when it enters/leaves scope. Another reason is that it is easier to see what is the type of an object if it is declared closer to the place where it is actually used. Finally, this is the more idiomatic way of coding c++.